src/main/scala/mesosphere/marathon/core/launchqueue/impl/ReviveOffersState.scala

```diff
@@ -5,83 +5,165 @@ import com.typesafe.scalalogging.StrictLogging
-    } else {
-      copy(scheduledInstances - instance.instanceId, terminalReservations - instance.instanceId)
-    }
+    updateInstanceState(instance.role, instance.instanceId, Some(instance))
   }
 
   /** @return this state with removed ref from [[activeDelays]]. */
   def withoutDelay(ref: RunSpecConfigRef): ReviveOffersState = {
     logger.debug(s"Marking $ref as no longer actively delayed")
-    copy(activeDelays = activeDelays - ref)
+
+    // This is not optimized
+    val bumpedVersions = hungryInstances.map {
+      case (role, hungryInstances) =>
+        role -> hungryInstances.map {
+          case (instanceId, hungryInstance) =>
+            if (hungryInstance.ref == ref)
+              instanceId -> hungryInstance.copy(version = this.version)
+            else
+              instanceId -> hungryInstance
+        }
+    }
+    update(hungryInstances = bumpedVersions, activeDelays = activeDelays - ref)
   }
 
   /** @return this state with updated [[activeDelays]]. */
   def withDelay(ref: RunSpecConfigRef): ReviveOffersState = {
     logger.debug(s"Marking $ref as actively delayed")
-    copy(activeDelays = activeDelays + ref)
+    update(activeDelays = activeDelays + ref)
   }
 
   /** scheduled instances that should be launched. */
-  lazy val scheduledInstancesWithoutBackoff: Set[Instance.Id] =
-    scheduledInstances.view.filter(kv => launchAllowed(kv._2)).map(_._1).toSet
+  lazy val roleReviveVersions: Map[String, RoleOfferState] = {
+    allInstanceIdsByRole.keysIterator.map { role =>
+      role -> hungryInstances.getOrElse(role, Map.empty).values
+        .iterator
+        .filter(launchAllowed)
+        .foldLeft(OffersNotWanted: RoleOfferState) {
+          case (op @ OffersWanted(version), hungryInstance) if hungryInstance.version < version =>
+            op
+          case (_, hungryInstance) =>
+            OffersWanted(hungryInstance.version)
+        }
+    }.toMap
+  }
 
   /** @return true if a instance has no active delay. */
-  def launchAllowed(ref: RunSpecConfigRef): Boolean = {
-    !activeDelays.contains(ref)
+  def launchAllowed(hungryInstance: HungryInstance): Boolean = {
+    hungryInstance.reason == ReviveReason.CleaningUpReservations || !activeDelays.contains(hungryInstance.ref)
   }
-
-  /** @return true there are no scheduled instances nor terminal instances with reservations. */
-  def isEmpty: Boolean = scheduledInstancesWithoutBackoff.isEmpty && terminalReservations.isEmpty
 }
 
 object ReviveOffersState {
-  def empty = ReviveOffersState(Map.empty, Set.empty, Set.empty, 0)
+  def empty = ReviveOffersState(Map.empty, Map.empty, Set.empty, 0)
+
+  case class HungryInstance(version: Long, reason: ReviveReason, ref: RunSpecConfigRef)
+
+  sealed trait ReviveReason
+
+  case object ReviveReason {
+
+    case object CleaningUpReservations extends ReviveReason
```

Comment:
I don't think we have this anymore, right?

Reply:
We do, instances can be in goal deccomissioned